File: sdk/java/src/main/java/run/runtime/sdk/RuntimeSubjects.java
package run.runtime.sdk;

/** Canonical runtime subject constructors shared by product facades. */
public final class RuntimeSubjects {
  static final String RUNTIME_STATE_READ_SUBJECT_PATH = "runtime-state/read";

  private RuntimeSubjects() {}

  /** Build the user-owned Resource URA used for runtime-state read projections. */
  public static String runtimeStateReadSubjectURA(String realm, String userID) {
    String cleanRealm = RuntimePrincipals.requiredString(realm, "realm", "runtime");
    String cleanUserID = RuntimePrincipals.requiredPrincipalID(userID, "user_id", "runtime");
    if (cleanRealm.contains("/") || cleanRealm.contains("?") || cleanRealm.contains("#")) {
      throw invalid("runtime-state read subject realm is not canonical");
    }
    if (cleanUserID.contains("/") || cleanUserID.contains("?") || cleanUserID.contains("#")) {
      throw invalid("runtime-state read subject user_id is not canonical");
    }
    String subject =
        "easynet:///r/"
            + cleanRealm
            + "/resource/user."
            + cleanUserID
            + "/"
            + RUNTIME_STATE_READ_SUBJECT_PATH;
    if (canonicalResourceSubject(subject) == null) {
      throw invalid("runtime-state read subject_ura must be canonical");
    }
    return subject;
  }

  static String descriptorBoundSubjectURA(String subjectURA, String abilityName) {
    String subject = RuntimePrincipals.requiredString(subjectURA, "subject_ura", "runtime");
    String ability = RuntimePrincipals.requiredString(abilityName, "ability name", "runtime");
    ParsedSubject parsed = parsedSubject(subject);
    if (parsed == null) {
      throw invalid("subject_ura is not a valid URA");
    }
    if ("authority".equals(parsed.path())) {
      return "easynet:///r/" + parsed.realm() + "/resource/authority/invoke/" + ability;
    }
    if (parsed.path().startsWith("user/")) {
      String userID = parsed.path().substring("user/".length()).trim();
      if (userID.isEmpty()
          || userID.contains("/")
          || userID.contains("?")
          || userID.contains("#")) {
        throw invalid("subject_ura user id is not canonical");
      }
      return "easynet:///r/" + parsed.realm() + "/resource/user." + userID + "/invoke/" + ability;
    }
    if (parsed.path().startsWith("agent/")
        || parsed.path().startsWith("ability/")
        || parsed.path().startsWith("device/")
        || parsed.path().startsWith("resource/")) {
      return subject;
    }
    throw invalid("subject_ura kind is not descriptor-bound");
  }

  static String runtimeGovernanceReadSubjectURA(String subjectURA) {
    String subject = RuntimePrincipals.requiredString(subjectURA, "subject_ura", "runtime");
    ResourceSubject resource = canonicalResourceSubject(subject);
    if (resource != null && RUNTIME_STATE_READ_SUBJECT_PATH.equals(resource.path())) {
      return subject;
    }
    ParsedSubject parsed = parsedSubject(subject);
    if (parsed != null && parsed.path().startsWith("user/")) {
      String userID = parsed.path().substring("user/".length()).trim();
      return runtimeStateReadSubjectURA(parsed.realm(), userID);
    }
    throw invalid("runtime governance read subject must be a User or user-owned runtime-state resource");
  }

  static ResourceSubject canonicalResourceSubject(String subjectURA) {
    if (subjectURA == null || RuntimePrincipals.containsAllZeroPrincipal(subjectURA)) {
      return null;
    }
    String raw = subjectURA.trim();
    String prefix = "easynet:///r/";
    if (!raw.startsWith(prefix)) {
      return null;
    }
    String rest = raw.substring(prefix.length());
    int slash = rest.indexOf('/');
    if (slash <= 0) {
      return null;
    }
    String path = rest.substring(slash + 1);
    String resourcePrefix = "resource/";
    if (!path.startsWith(resourcePrefix)) {
      return null;
    }
    String resource = path.substring(resourcePrefix.length());
    int pathSlash = resource.indexOf('/');
    if (pathSlash <= 0 || pathSlash == resource.length() - 1) {
      return null;
    }
    String ownerID = resource.substring(0, pathSlash).trim();
    String resourcePath = resource.substring(pathSlash + 1).trim();
    if (ownerID.isEmpty()
        || ownerID.contains("/")
        || resourcePath.isEmpty()
        || resourcePath.startsWith("/")
        || resourcePath.contains("//")) {
      return null;
    }
    return new ResourceSubject(ownerID, resourcePath);
  }

  static boolean isRuntimeGovernanceReadSubject(String subjectURA, String calleeURA) {
    String subject = subjectURA == null ? "" : subjectURA.trim();
    if (subject.isEmpty() || RuntimePrincipals.containsAllZeroPrincipal(subject)) {
      return false;
    }
    ResourceSubject resource = canonicalResourceSubject(subject);
    if (resource != null) {
      String userOwner = resource.ownerID().startsWith("user.")
          ? resource.ownerID().substring("user.".length()).trim()
          : "";
      return !userOwner.isEmpty()
          && !userOwner.contains(".")
          && !RuntimePrincipals.containsAllZeroPrincipal(userOwner)
          && RUNTIME_STATE_READ_SUBJECT_PATH.equals(resource.path());
    }
    RuntimeOwnerSubject parsedSubject = runtimeOwnerSubject(subject);
    RuntimeOwnerSubject parsedCallee = runtimeOwnerSubject(calleeURA);
    return parsedSubject != null
        && parsedCallee != null
        && parsedSubject.kind().equals(parsedCallee.kind())
        && parsedSubject.realm().equals(parsedCallee.realm())
        && subject.equals(calleeURA == null ? "" : calleeURA.trim());
  }

  private static RuntimeOwnerSubject runtimeOwnerSubject(String ura) {
    ParsedSubject parsed = parsedSubject(ura);
    if (parsed == null) {
      return null;
    }
    String realm = parsed.realm();
    String path = parsed.path();
    if (realm.isEmpty() || realm.contains("/")) {
      return null;
    }
    if (path.startsWith("authority") && path.equals("authority")) {
      return new RuntimeOwnerSubject("authority", realm);
    }
    if (path.startsWith("device/")) {
      String deviceID = path.substring("device/".length()).trim();
      if (deviceID.isEmpty() || deviceID.contains("/")) {
        return null;
      }
      return new RuntimeOwnerSubject("device", realm);
    }
    return null;
  }

  private static ParsedSubject parsedSubject(String ura) {
    String raw = ura == null ? "" : ura.trim();
    String prefix = "easynet:///r/";
    if (!raw.startsWith(prefix) || RuntimePrincipals.containsAllZeroPrincipal(raw)) {
      return null;
    }
    String rest = raw.substring(prefix.length());
    int slash = rest.indexOf('/');
    if (slash <= 0 || slash == rest.length() - 1) {
      return null;
    }
    String realm = rest.substring(0, slash).trim();
    String path = rest.substring(slash + 1).trim();
    if (realm.isEmpty() || realm.contains("/") || path.isEmpty()) {
      return null;
    }
    return new ParsedSubject(realm, path);
  }

  static boolean canonicalSessionAuthorityID(String sessionID) {
    if (sessionID == null) {
      return false;
    }
    String cleaned = sessionID.trim();
    if (cleaned.isEmpty()) {
      return false;
    }
    for (int index = 0; index < cleaned.length(); index++) {
      char ch = cleaned.charAt(index);
      if ((ch >= 'a' && ch <= 'z')
          || (ch >= 'A' && ch <= 'Z')
          || (ch >= '0' && ch <= '9')
          || ch == '-'
          || ch == '.') {
        continue;
      }
      return false;
    }
    return true;
  }

  private static SDKError invalid(String message) {
    return SDKError.validation("runtime", message);
  }

  record ResourceSubject(String ownerID, String path) {}

  private record RuntimeOwnerSubject(String kind, String realm) {}

  private record ParsedSubject(String realm, String path) {}
}

File: sdk/swift/Sources/RuntimeSDK/RuntimeSubjects.swift
import Foundation

enum RuntimeSubjects {
    private static let runtimeStateReadSubjectPath = "runtime-state/read"

    static func runtimeStateReadSubjectURA(realm: String, userID: String) throws -> String {
        let cleanRealm = try RuntimePrincipals.requiredString(realm, "realm", stage: "runtime")
        let cleanUserID = try RuntimePrincipals.requiredPrincipalID(userID, "user_id", stage: "runtime")
        guard !cleanRealm.contains("/"), !cleanRealm.contains("?"), !cleanRealm.contains("#") else {
            throw invalidRuntime("runtime-state read subject realm is not canonical")
        }
        guard !cleanUserID.contains("/"), !cleanUserID.contains("?"), !cleanUserID.contains("#") else {
            throw invalidRuntime("runtime-state read subject user_id is not canonical")
        }
        let subject = "easynet:///r/\(cleanRealm)/resource/user.\(cleanUserID)/\(runtimeStateReadSubjectPath)"
        guard canonicalResourceSubject(subject) != nil else {
            throw invalidRuntime("runtime-state read subject_ura must be canonical")
        }
        return subject
    }

    static func descriptorBoundSubjectURA(_ subjectURA: String, abilityName: String) throws -> String {
        let subject = try RuntimePrincipals.requiredString(subjectURA, "subject_ura", stage: "runtime")
        let ability = try RuntimePrincipals.requiredString(abilityName, "ability name", stage: "runtime")
        guard let parsed = parsedSubject(subject) else {
            throw invalidRuntime("subject_ura is not a valid URA")
        }
        if parsed.path == "authority" {
            return "easynet:///r/\(parsed.realm)/resource/authority/invoke/\(ability)"
        }
        let userPrefix = "user/"
        if parsed.path.hasPrefix(userPrefix) {
            let userID = String(parsed.path.dropFirst(userPrefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
            guard !userID.isEmpty,
                  !userID.contains("/"),
                  !userID.contains("?"),
                  !userID.contains("#")
            else {
                throw invalidRuntime("subject_ura user id is not canonical")
            }
            return "easynet:///r/\(parsed.realm)/resource/user.\(userID)/invoke/\(ability)"
        }
        if parsed.path.hasPrefix("agent/") ||
            parsed.path.hasPrefix("ability/") ||
            parsed.path.hasPrefix("device/") ||
            parsed.path.hasPrefix("resource/")
        {
            return subject
        }
        throw invalidRuntime("subject_ura kind is not descriptor-bound")
    }

    static func canonicalResourceSubject(_ subjectURA: String) -> ResourceSubject? {
        guard !RuntimePrincipals.containsAllZeroPrincipal(subjectURA) else {
            return nil
        }
        let raw = subjectURA.trimmingCharacters(in: .whitespacesAndNewlines)
        let realmPrefix = "easynet:///r/"
        guard raw.hasPrefix(realmPrefix) else {
            return nil
        }
        let rest = String(raw.dropFirst(realmPrefix.count))
        guard let slash = rest.firstIndex(of: "/"), slash != rest.startIndex else {
            return nil
        }
        let path = String(rest[rest.index(after: slash)...])
        let resourcePrefix = "resource/"
        guard path.hasPrefix(resourcePrefix) else {
            return nil
        }
        let resource = String(path.dropFirst(resourcePrefix.count))
        guard let pathSlash = resource.firstIndex(of: "/"),
              pathSlash != resource.startIndex,
              pathSlash != resource.index(before: resource.endIndex)
        else {
            return nil
        }
        let ownerID = String(resource[..<pathSlash]).trimmingCharacters(in: .whitespacesAndNewlines)
        let resourcePath = String(resource[resource.index(after: pathSlash)...]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !ownerID.isEmpty,
              !ownerID.contains("/"),
              !resourcePath.isEmpty,
              !resourcePath.hasPrefix("/"),
              !resourcePath.contains("//")
        else {
            return nil
        }
        return ResourceSubject(ownerID: ownerID, path: resourcePath)
    }

    static func isRuntimeGovernanceReadSubject(_ subjectURA: String, calleeURA: String) -> Bool {
        let subject = subjectURA.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !subject.isEmpty, !RuntimePrincipals.containsAllZeroPrincipal(subject) else {
            return false
        }
        if let resource = canonicalResourceSubject(subject) {
            let userPrefix = "user."
            guard resource.ownerID.hasPrefix(userPrefix) else {
                return false
            }
            let userID = String(resource.ownerID.dropFirst(userPrefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
            return !userID.isEmpty &&
                !userID.contains(".") &&
                !RuntimePrincipals.containsAllZeroPrincipal(userID) &&
                resource.path == runtimeStateReadSubjectPath
        }
        guard let parsedSubject = runtimeOwnerSubject(subject),
              let parsedCallee = runtimeOwnerSubject(calleeURA)
        else {
            return false
        }
        return parsedSubject.kind == parsedCallee.kind &&
            parsedSubject.realm == parsedCallee.realm &&
            subject == calleeURA.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func runtimeOwnerSubject(_ ura: String) -> RuntimeOwnerSubject? {
        guard let parsed = parsedSubject(ura) else { return nil }
        let realm = parsed.realm
        let path = parsed.path
        guard !realm.isEmpty, !realm.contains("/") else {
            return nil
        }
        if path == "authority" {
            return RuntimeOwnerSubject(kind: "authority", realm: realm)
        }
        let devicePrefix = "device/"
        if path.hasPrefix(devicePrefix) {
            let deviceID = String(path.dropFirst(devicePrefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
            guard !deviceID.isEmpty, !deviceID.contains("/") else {
                return nil
            }
            return RuntimeOwnerSubject(kind: "device", realm: realm)
        }
        return nil
    }

    private static func parsedSubject(_ ura: String) -> ParsedSubject? {
        let raw = ura.trimmingCharacters(in: .whitespacesAndNewlines)
        let realmPrefix = "easynet:///r/"
        guard raw.hasPrefix(realmPrefix), !RuntimePrincipals.containsAllZeroPrincipal(raw) else {
            return nil
        }
        let rest = String(raw.dropFirst(realmPrefix.count))
        guard let slash = rest.firstIndex(of: "/"),
              slash != rest.startIndex,
              slash != rest.index(before: rest.endIndex)
        else {
            return nil
        }
        let realm = String(rest[..<slash]).trimmingCharacters(in: .whitespacesAndNewlines)
        let path = String(rest[rest.index(after: slash)...]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !realm.isEmpty, !realm.contains("/"), !path.isEmpty else {
            return nil
        }
        return ParsedSubject(realm: realm, path: path)
    }

    static func canonicalSessionAuthorityID(_ sessionID: String) -> Bool {
        let cleaned = sessionID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else {
            return false
        }
        return cleaned.unicodeScalars.allSatisfy { scalar in
            let value = scalar.value
            return (value >= 0x61 && value <= 0x7A) ||
                (value >= 0x41 && value <= 0x5A) ||
                (value >= 0x30 && value <= 0x39) ||
                value == 0x2D ||
                value == 0x2E
        }
    }

    struct ResourceSubject {
        let ownerID: String
        let path: String
    }

    private struct RuntimeOwnerSubject {
        let kind: String
        let realm: String
    }

    private struct ParsedSubject {
        let realm: String
        let path: String
    }

    private static func invalidRuntime(_ message: String) -> SDKError {
        SDKError.validation("runtime", message)
    }
}
